File: Repository/src/main/java/org/example/EventDBRepository.java
package org.example;

import Utils.JdbcUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.*;

public class EventDBRepository implements EventRepository {

    private final JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();

    public EventDBRepository(Properties props) {
        logger.info("Initializing EventDBRepository with properties: {} ", props);
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public void add(Event entity) throws EntityRepoException {
        logger.traceEntry("add task {} ", entity);
        Connection connection = dbUtils.getConnection();
        String sql = "INSERT INTO Events(style, distance) values (?,?)";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setString(1, entity.getStyle());
            ps.setInt(2, entity.getDistance());
            ps.executeUpdate();
            logger.traceExit("task {} added", entity);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
            throw new EntityRepoException(e);
        }
        logger.traceExit("task {} added", entity);
    }

    @Override
    public void remove(long id) throws EntityRepoException {
        logger.traceEntry("remove task {} ", id);
        Connection connection = dbUtils.getConnection();
        String sql = "DELETE FROM Events WHERE id=?";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setLong(1, id);
            ps.executeUpdate();
            logger.traceExit("task {} removed", id);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
            throw new EntityRepoException(e);
        }
        logger.traceExit("task {} removed", id);
    }

    @Override
    public List<Event> getAll() throws EntityRepoException {
        logger.traceEntry("task getALL");
        Connection connection = dbUtils.getConnection();
        List<Event> events = new ArrayList<>();
        String sql = "SELECT * FROM Events";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    Event event = extract(rs);
                    events.add(event);
                    logger.traceExit("task {} added", event);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
            throw new EntityRepoException(e);
        }
        return events;
    }

    @Override
    public Event findById(long id) throws EntityRepoException {
        logger.traceEntry("task findById {}", id);
        Connection connection = dbUtils.getConnection();
        String sql = "SELECT * FROM Events WHERE id=?";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setLong(1, id);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    Event event = extract(rs);
                    logger.traceExit("task {} found", event);
                    return event;
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
            throw new EntityRepoException(e);
        }
        return null;
    }

    @Override
    public void update(long id, Event entity) throws EntityRepoException {
        logger.traceEntry("update task {} ", entity);
        Connection connection = dbUtils.getConnection();
        String sql = "UPdate Events SET style=?, distance=? WHERE id=?";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setString(1, entity.getStyle());
            ps.setInt(2, entity.getDistance());
            ps.setLong(3, entity.getId());
            ps.executeUpdate();
            logger.traceExit("task {} updated", entity);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
            throw new EntityRepoException(e);
        }
    }

    private Event extract(ResultSet resultSet) throws SQLException {
        Long id = resultSet.getLong("id");
        String style = resultSet.getString("style");
        int distance = resultSet.getInt("distance");
        Event event = new Event(style, distance);
        event.setId(id);
        return event;
    }
}

File: java-app/Repository/EventDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using log4net;
using mpp_proiect_csharp_DianaGliga11.Model;
using mpp_proiect_csharp_DianaGliga11.Repository;

namespace mpp_proiect_csharp_DianaGliga11.Repository
{
    public class EventDBRepository : I_EventDBRepository
    {
        private static readonly ILog log = LogManager.GetLogger(typeof(EventDBRepository));
        private readonly IDictionary<string, string> Props;
        
        public EventDBRepository(IDictionary<string, string> props) 
        {
            log.Info($"{nameof(EventDBRepository)} constructed.");
            Props = props;
        }

        public void Add(Event entity)
        {
            log.Info($"Adding Event: {entity}");
            var connection = DBUtils.GetConnection(Props);

            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = @"INSERT INTO Events (style, distance) VALUES (@style, @distance)";

                var styleParam = command.CreateParameter();
                styleParam.ParameterName = "@style";
                styleParam.Value = entity.Style;
                command.Parameters.Add(styleParam);

                var distanceParam = command.CreateParameter();
                distanceParam.ParameterName = "@distance";
                distanceParam.Value = entity.Distance;
                command.Parameters.Add(distanceParam);

                var result = command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                log.Error("Error while adding new Event ", ex);
                throw new EntityRepoException(ex);
            }

            log.Info($"Added Event: {entity}");
        }

        public void Remove(long id)
        {
            log.Info($"Removing Event: {id}");
            var connection = DBUtils.GetConnection(Props);
            
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = @"DELETE FROM Events WHERE id = @id";
                var idParam = command.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = id;
                command.Parameters.Add(idParam);
                var result = command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                log.Error("Error while removing Event ", ex);
                throw new EntityRepoException(ex);
            }
            
            log.Info($"Removed Event: {id}");
        }

        public void Update(long id, Event entity)
        {
            log.Info($"Updating Event: {entity}");
            var connection = DBUtils.GetConnection(Props);
            
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = "UPDATE Events SET style = @style, distance = @distance WHERE id = @id";

                var styleParam = command.CreateParameter();
                styleParam.ParameterName = "@style";
                styleParam.Value = entity.Style;
                command.Parameters.Add(styleParam);

                var distanceParam = command.CreateParameter();
                distanceParam.ParameterName = "@distance";
                distanceParam.Value = entity.Distance;
                command.Parameters.Add(distanceParam);

                var idParam = command.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = id;
                command.Parameters.Add(idParam);

                var result = command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                log.Error("Error while updating Event ", ex);
                throw new EntityRepoException(ex);
            }
            
            log.Info($"Updated Event: {entity}");
        }
        
        public Event findById(long id)
        {
            log.Info($"Finding Event: {id}");
            var connection = DBUtils.GetConnection(Props);
            
            try
            {
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "SELECT * FROM Events WHERE id = @id";
                    var paramID = command.CreateParameter();
                    paramID.ParameterName = "@id";
                    paramID.Value = id;
                    command.Parameters.Add(paramID);

                    using (var dataReader = command.ExecuteReader())
                    {
                        if (dataReader.Read())
                        {
                            Event evt = Extract(dataReader);
                            log.Info($"Found Event: {evt}");
                            return evt;
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                log.Error("Error while finding Event ", ex);
                throw new EntityRepoException("Event was not found");
            }

            return null;
        }

        public IEnumerable<Event> getAll()
        {
            log.Info("Getting All Events");
            var connection = DBUtils.GetConnection(Props);
            IList<Event> events = new List<Event>();
            
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "SELECT * FROM Events";
                
                using (var dataReader = command.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        Event evt = Extract(dataReader);
                        events.Add(evt);
                    }
                }
            }
            
            log.Info($"Retrieved {events.Count} events");
            return events;
        }
        

        private Event Extract(IDataReader dataReader)
        {
            long id = dataReader.GetInt64(0);
            string style = dataReader.GetString(1);
            int distance = dataReader.GetInt32(2);
            Event evt = new Event(style, distance)
            {
                Id = id
            };

            return evt;
        }
    }
}